File: 99questions/1to10.hs
-- 1 find the last element of a list
-- > myLast [1,2,3,4]
-- 4
-- > myLast ['x', 'y', 'z']
-- 'z'
myLast :: [a] -> a
myLast [] = error "No last element of empty list"
myLast [x] = x
myLast (_: xs) = myLast xs

-- 2 find the last but one element of a list
myButLast :: [a] -> a
myButLast [] = error "No last but one element of empty list"
myButLast [x] = error "No last but of one element list"
myButLast [x, _] = x
myButLast (_:xs) = myButLast xs

-- 3 find the kth element of a list. the first element is number 1
elementAt :: [a] -> Integer -> a
elementAt (x: _) 1 = x
elementAt [] _ = error "cant do empty list"
elementAt (_:xs) k
  | k < 1 = error "k too small"
  | otherwise = elementAt xs (k-1)

elementAt' xs n
  | length xs < n = error "Index out of bounds"
  | otherwise = fst . last $ zip xs [1..n]

-- 4 find the number of elements in a list
myLength :: [a] -> Integer
myLength [] = 0
myLength (_:xs) = 1 + myLength xs

File: 99questions/1to10.js
// 1 find the last element of a list
const myLast = (list) => {
  const [x, y, ...z] = list;
  // list.length === 0
  if(x === undefined) {
    console.error("sorry empty array!");
    return null;
  }
  // list.length === 1 || list.length === 2
  if(z.length === 0) {
    return y || x;
  }
  // list.length > 2
  return myLast(list.slice(1));
};

// 2 find the last but one element of a list
const myButLast = (list) => {
  const [x, ...y] = list;
  if(x === undefined) {
    console.error("sorry empty array!");
    return null;
  }
  if(y.length === 0) {
    console.error("sorry array too short");
    return null;
  }
  if(y.length === 1) {
    return x;
  }
  return myButLast(list.slice(1));
};

// 3 find the kth element of a list. the first element is number 1
const elementAt = (list, k) => {
  const [x, y] = list;
  if (k === 1) {
    return x;
  }
  return elementAt(y, k-1);
};
